lib/Sema/TypeCheckProtocol.cpp

```diff
@@ -530,7 +530,114 @@ swift::matchWitness(
   }
 
   // Now finalize the match.
-  return finalize(anyRenaming, optionalAdjustments);
+  auto result = finalize(anyRenaming, optionalAdjustments);
```

Comment:
Note: I anticipate that maybe some code owners here would care strongly about preserving `return finalize(anyRenaming, optionalAdjustments);`. It doesn't matter too much to me, so I upstreamed code from `tensorflow` branch as in.

Aside: this logic for diagnosing unmet `@differentiable` attribute requirement doesn't work when associated type inference is involved in witness checking. TF-1014 tracks that issue.

Reply:
> Note: I anticipate that some code owners might care strongly about preserving `return finalize(anyRenaming, optionalAdjustments)` here.

Slightly mitigating the above point: I refactored `swift::matchWitness` logic for checking `@differentiable` attributes into a `matchWitnessDifferentiableAttr` helper. So the code changes to `swift::matchWitness` are minimal.